apollo-federation/src/sources/connect/models/http_json_transport.rs

```diff
@@ -0,0 +1,321 @@
+use std::error::Error;
+use std::fmt::Display;
+use std::fmt::Formatter;
+use std::str::FromStr;
+
+use apollo_compiler::Node;
+use apollo_compiler::ast;
+use apollo_compiler::collections::IndexMap;
+use apollo_compiler::parser::SourceSpan;
+use either::Either;
+use http::HeaderName;
+use http::Uri;
+
+use crate::error::FederationError;
+use crate::sources::connect::JSONSelection;
+use crate::sources::connect::Namespace;
+use crate::sources::connect::PathSelection;
+use crate::sources::connect::StringTemplate;
+use crate::sources::connect::header::HeaderValue;
+use crate::sources::connect::json_selection::ExternalVarPaths;
+use crate::sources::connect::spec::ConnectHTTPArguments;
+use crate::sources::connect::spec::SourceHTTPArguments;
+use crate::sources::connect::spec::schema::HEADERS_ARGUMENT_NAME;
+use crate::sources::connect::spec::schema::HTTP_HEADER_MAPPING_FROM_ARGUMENT_NAME;
+use crate::sources::connect::spec::schema::HTTP_HEADER_MAPPING_NAME_ARGUMENT_NAME;
+use crate::sources::connect::spec::schema::HTTP_HEADER_MAPPING_VALUE_ARGUMENT_NAME;
+use crate::sources::connect::spec::versions::AllowedHeaders;
+use crate::sources::connect::string_template;
+use crate::sources::connect::variable::VariableReference;
+
+#[derive(Clone, Debug)]
+pub struct HttpJsonTransport {
+    pub source_url: Option<Uri>,
+    pub connect_template: StringTemplate,
+    pub method: HTTPMethod,
+    pub headers: IndexMap<HeaderName, HeaderSource>,
+    pub body: Option<JSONSelection>,
+}
+
+impl HttpJsonTransport {
+    pub(crate) fn from_directive(
+        http: &ConnectHTTPArguments,
+        source: Option<&SourceHTTPArguments>,
+    ) -> Result<Self, FederationError> {
+        let (method, connect_url) = if let Some(url) = &http.get {
+            (HTTPMethod::Get, url)
+        } else if let Some(url) = &http.post {
+            (HTTPMethod::Post, url)
+        } else if let Some(url) = &http.patch {
+            (HTTPMethod::Patch, url)
+        } else if let Some(url) = &http.put {
+            (HTTPMethod::Put, url)
+        } else if let Some(url) = &http.delete {
+            (HTTPMethod::Delete, url)
+        } else {
+            return Err(FederationError::internal("missing http method"));
+        };
+
+        #[allow(clippy::mutable_key_type)]
+        // HeaderName is internally mutable, but we don't mutate it
```

Comment:
I know this was already here, but if we don't need mutability and we only need to preserve the headers from `http.headers`  as the source of truth, we can prevent the early headers cloning by doing:

```rus
let headers = source.map(|source| &source.headers)
            .into_iter()
            .flatten()
            .chain(http.headers.iter())
            .map(|(name, souce)| (name.clone(), souce.clone()))
            .collect::<IndexMap<HeaderName, HeaderSource>>();
```